ifupdown/ifupdownmain.py

```diff
@@ -607,6 +607,32 @@ def query_upperifaces(self, ifaceobj, ops, ifacenames, type=None):
             if ulist: ret_ulist.extend(ulist)
         return list(set(ret_ulist))
 
+    def _remove_circular_veth_dependencies (self, ifaceobj, dlist):
+        # if ifaceobj isn't a veth link, ignore it.
+        if ifaceobj.get_attr_value_first('link-type') != "veth":
+            return
+
+        for diface in dlist:
+            difaceobj = self.get_ifaceobj_first(diface)
+            # If the dependent iface isn't a veth link - which shouldn't
+            # happen - ignore it to be save.
+            if difaceobj and difaceobj.get_attr_value_first('link-type') != "veth":
+                continue
+
+            # If the peer has a desired peer name set and this is us,
+            # see if the peer has a dependency to us too and remove our
+            # redundant dependency to the peer.
+            diface_peer_name = difaceobj.get_attr_value_first('veth-peer-name')
+            if diface_peer_name and diface_peer_name == ifaceobj.name:
+                peer_dlist = difaceobj.lowerifaces
+                if not peer_dlist:
+                    # Not list of dependent interface on the peer.
+                    continue
+
+                # We aleady are in the peers dlist, don't add dependcy from us to peer
+                if ifaceobj.name in peer_dlist:
+                    dlist.remove (difaceobj.name)
+
```

Comment:
this is a nice trick to solving the circular dependency. But, I would prefer link specific code not bubble up into ifupdownmain. In the most common use case, the other end of the veth peer is connected to a separate network namespace or container and managed separately. ifupdown2 does not support namespaces today.
I would prefer to not bring in this circular dependency handing in ifupdownmain if that can be avoided.
Assuming you are using this in the same network namespace, will the below not work for you ?.

auto veth-A
iface veth-A
      link-type veth
      link-veth-peer-name veth-B

auto veth-B
iface veth-B
       
veth-B will show up as a dependency of veth-A, and hence veth-A will always be created before veth-B...and everything should fall through fine after that. correct ?. are you seeing problems with this ?.